Mark complete button

hynesy23 opened the pull request.

code-mattclaffey commented on _site/_includes/lessons/01-removing-critical-chain-requests.njk:
```diff
@@ -0,0 +1,26 @@
+<!doctype html>
```
Can we change this folder to match exercises?

code-mattclaffey commented on _site/assets/js/global/mark-completed.js:
```diff
@@ -1,15 +1,151 @@
+  };
+
+  const getSections = () => {
+    const sections = document.querySelectorAll('[data-section]');
```
These can be declared at the top of the function once and used everywhere.

code-mattclaffey commented on _site/assets/js/global/mark-completed.js:
```diff
@@ -1,15 +1,151 @@
+  };
+
+  const getButtons = () => {
+    const buttons = document.querySelectorAll('[data-mark-complete-btn]');
```
same as above

code-mattclaffey commented on _site/assets/js/global/mark-completed.js:
```diff
@@ -1,15 +1,151 @@
+
+  const updateButton = (btn, action) => {
+    if (action === 'remove') {
+      btn.classList.remove('bg-secondary');
```
You can do something easier here. 

```
const originalClassNames = btn.getAttribute('class') // get the initial class names
btn.removeAttribute('class');
btn.setAttribute('class', 'bg-primary text-secondary hover:bg-secondary so on so on')
```

code-mattclaffey replied:
You can even do a ternary for this as well.

code-mattclaffey commented on _site/assets/js/global/mark-completed.js:
```diff
@@ -1,15 +1,151 @@
+    }
+  };
+
+  init();
```
You have clearly done this before :eyes:

code-mattclaffey commented on _site/assets/js/global/mark-completed.js:
```diff
@@ -1,15 +1,151 @@
+  const handleOnClick = (event) => {
+    const target = event.target;
+    if (target.dataset.completed === 'false') {
+      updateLocalStorage(target, 'add');
```
const action = target.dataset.completed === 'false' ? 'add' : 'remove'

 updateLocalStorage(target, action);
      updateButton(target, action);

code-mattclaffey commented on _site/assets/js/global/mark-completed.js:
```diff
@@ -1,15 +1,151 @@
+  };
+
+  const getStoreFromLocalStorage = () => {
+    const store = JSON.parse(localStorage.getItem('performanceKit')) || {};
```
You can call this up top too

code-mattclaffey commented on _site/assets/js/global/mark-completed.js:
```diff
@@ -1,15 +1,151 @@
+
+  const addCompleteFlagToSections = () => {
+    const sections = getSections();
+    const completedLessons = getStoreFromLocalStorage().completedLessons;
```
Can do a cheeky `??` here where you do 

const completedLessons = getStoreFromLocalStorage().completedLessons ?? [];

also with the looping, it's a bit mental. You do this instead:

```
sections.forEach(section => {
  const isCompleted = completedLessons.find(lesson => lesson === section.id);

  if (isCompleted) {
    section.classList.add('completed');
    section.querySelector('[data-complete-tag]').classList.remove('invisible');
  } else {
     section.classList.remove('completed');
     section.querySelector('[data-complete-tag]').classList.add('invisible');
  }
});
```

code-mattclaffey commented on _site/assets/js/global/mark-completed.js:
```diff
@@ -1,15 +1,151 @@
+    const buttons = getButtons();
+    const completedLessons = getStoreFromLocalStorage().completedLessons;
+
+    if (completedLessons && completedLessons.length) {
```
Same as below.

code-mattclaffey commented on _site/assets/js/global/mark-completed.js:
```diff
@@ -1,15 +1,151 @@
+  const assignOnClickToButtons = () => {
+    const buttons = getButtons();
+
+    for (let i = 0; i < buttons.length; i++) {
```
Nicer to change this to a buttons.forEach()

code-mattclaffey commented on _site/_includes/components/_mark-complete-btn.njk:
```diff
@@ -0,0 +1,3 @@
+<div class="text-center">
+    <button data-postId="{{ postId }}" data-mark-complete-btn data-completed="false" class="w-2/5 p-4 bg-primary text-secondary rounded hover:bg-secondary hover:text-primary" id="mark-complete">Mark lesson complete</button>
```
'block max-w-xs'

Commit:
Implement mark-complete button

Implemented ability to mark work complete  in content pages so people know how much progres they've made/what they've completed
